File: editors/mcp-client/src/masking.ts
export function maskValue(value: string, maskChar = "*"): string {
  if (!value || value.length === 0) {
    return "";
  }
  // Always show 3 mask chars regardless of value length
  return maskChar.repeat(3);
}

/**
 * Recursively mask all values in an object while preserving structure.
 * Keys like 'path', 'meta', 'type' are preserved unmasked.
 */
export function maskObject(
  obj: Record<string, unknown>,
  preserveKeys: string[] = ["path", "meta", "type", "name", "version", "created", "updated"]
): Record<string, unknown> {
  const result: Record<string, unknown> = {};

  for (const [key, value] of Object.entries(obj)) {
    if (preserveKeys.includes(key)) {
      result[key] = value;
    } else if (typeof value === "string") {
      result[key] = maskValue(value);
    } else if (typeof value === "object" && value !== null) {
      result[key] = maskObject(value as Record<string, unknown>, preserveKeys);
    } else {
      result[key] = value;
    }
  }

  return result;
}

/**
 * Mask an Entry's data fields while preserving metadata.
 */
export function maskEntryData(entry: {
  path: string;
  data: Record<string, unknown>;
  meta: unknown;
}): { path: string; data: Record<string, unknown>; meta: unknown } {
  return {
    path: entry.path,
    data: maskObject(entry.data),
    meta: entry.meta,
  };
}

File: editors/nvim/lua/symaira/masking.lua
local config = require("symvault.config")

local M = {}

--- Mask a single value, showing only mask characters.
---@param value string
---@return string
function M.mask_value(value)
  if not value or value == "" then
    return ""
  end
  local mask_char = config.get("masking_char")
  return string.rep(mask_char, 3)
end

--- Check if a key should be preserved (not masked).
---@param key string
---@return boolean
local function is_preserved(key)
  local preserved = config.get("preserve_keys")
  for _, k in ipairs(preserved) do
    if k == key then
      return true
    end
  end
  return false
end

--- Recursively mask all string values in a table while preserving structure.
--- Non-string values (numbers, booleans, nil) are kept as-is.
--- Keys in preserve_keys are left unmasked.
---@param obj table
---@return table
function M.mask_object(obj)
  if type(obj) ~= "table" then
    return obj
  end
  local result = {}
  for k, v in pairs(obj) do
    if is_preserved(k) then
      result[k] = v
    elseif type(v) == "string" then
      result[k] = M.mask_value(v)
    elseif type(v) == "table" then
      result[k] = M.mask_object(v)
    else
      result[k] = v
    end
  end
  return result
end

--- Mask an entry's data fields while preserving path and metadata.
---@param entry table Entry with path, data, and meta fields.
---@return table Masked copy of the entry.
function M.mask_entry(entry)
  if not entry then
    return nil
  end
  local masked = {}
  masked.path = entry.path
  masked.data = M.mask_object(entry.data or {})
  masked.meta = entry.meta
  return masked
end

return M
